src/node_crypto.cc

```diff
@@ -685,26 +728,24 @@ void SecureContext::AddCACert(const FunctionCallbackInfo<Value>& args) {
     return env->ThrowTypeError("CA certificate argument is mandatory");
   }
 
-  if (!sc->ca_store_) {
-    sc->ca_store_ = X509_STORE_new();
-    newCAStore = true;
+  BIO* bio = LoadBIO(env, args[0]);
+  if (!bio) {
+    return;
   }
 
-  unsigned cert_count = 0;
-  if (BIO* bio = LoadBIO(env, args[0])) {
-    while (X509* x509 =
-        PEM_read_bio_X509(bio, nullptr, CryptoPemCallback, nullptr)) {
-      X509_STORE_add_cert(sc->ca_store_, x509);
-      SSL_CTX_add_client_CA(sc->ctx_, x509);
-      X509_free(x509);
-      cert_count += 1;
+  X509_STORE* cert_store = SSL_CTX_get_cert_store(sc->ctx_);
+  while (X509* x509 =
+             PEM_read_bio_X509(bio, nullptr, CryptoPemCallback, nullptr)) {
+    if (cert_store == root_cert_store) {
+      cert_store = NewRootCertStore();
```

Comment:
> If the existing store is root_cert_store (i.e. the global root store) then it'll be replaced with a local copy of the same root store. Otherwise modifying the store would change it for every context.

I understand the bug you fixed, but you changed the behaviour. We need to fix the bug wherein the root store gets modified - because the root store should *never* be modified. That was good work.

However, unless I am very much mistaken, you introduced an unintended side-effect at the same time. All use of the `ca:` option is intended to remove the root CAs, and replace them (for this specific secure context). For better or worse, that is the documented behaviour and pre-existing behaviour.

I would like to add  a unit test to prevent regressions here, but its quite difficult, because I need an identity signed by one of the CAs in our default roots, and to prove that using the `ca:` option makes that identity no longer verifiable. Or, I could make an outgoing connection to google.com, and prove that once an explict `ca:` is set, I can't verify the cert, but the network-required tests are annoying. Still, maybe less annoying than getting a cert issued to me.

@shigeki do you have any ideas how I would approach this?